src/storage/lsm.rs

```diff
@@ -119,65 +127,107 @@ impl SSTable {
+        let sstables: Vec<SSTable> = std::fs::read_dir(SSTABLES_DIR_PATH)?
+            .map(|dir_iter| dir_iter.map(|e| e.path()))
             .map(|path| SSTable::read_from_file(path.unwrap()).unwrap())
             .collect();
 
         let ret = State {
             memtable,
-            commit_log: Some(commit_log),
+            commit_log_path,
+            commit_log,
+            memtable_in_flush: None,
             sstables,
         };
         Ok(ret)
     }
 
-    pub fn flush_memtable(&mut self) -> Result<()> {
-        let path = PathBuf::from(format!(
-            "/tmp/pancake/sstables/{}.data",
-            SystemTime::now()
-                .duration_since(UNIX_EPOCH)
-                .unwrap()
-                .as_micros()
-        ));
-        let sstable = SSTable::write_from_memtable(&self.memtable, path)?;
-        self.sstables.push(sstable);
-        self.memtable.0.clear();
-        self.commit_log.take(); // Close the file.
-        self.commit_log = Some(
-            OpenOptions::new()
-                .create(true)
-                .write(true)
-                .append(true)
-                .open(COMMIT_LOG_PATH)?,
-        );
+    fn flush_memtable(&mut self) -> Result<()> {
+        let new_cl_path = new_path(COMMIT_LOGS_DIR_PATH);
+        let new_cl = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .append(true)
+            .open(&new_cl_path)?;
+        let old_cl_path: PathBuf;
+        {
+            // TODO MutexGuard here
+            let old_mt = mem::replace(&mut self.memtable, Memtable::default());
+            self.memtable_in_flush = Some(old_mt);
+
+            self.commit_log = new_cl;
+            old_cl_path = mem::replace(&mut self.commit_log_path, new_cl_path);
+        }
+
+        let new_sst = SSTable::write_from_memtable(self.memtable_in_flush.as_ref().unwrap(), new_path(SSTABLES_DIR_PATH))?;
+        {
+            // TODO MutexGuard here
+            self.sstables.push(new_sst);
+            self.memtable_in_flush.take();
+        }
+        println!("Removing {:?}", &old_cl_path);
+        fs::remove_file(old_cl_path)?;
 
         if self.sstables.len() >= SSTABLE_COMPACT_COUNT_THRESH {
-            // TODO compact
+            self.compact_sstables()?;
+        }
+
+        Ok(())
+    }
+
+    fn compact_sstables(&mut self) -> Result<()> {
+        let mut dense_idx = Memtable::default();
+        for old_sst in self.sstables.iter() {
+            dense_idx.update_from_commit_log(&old_sst.path)?;
+        }
+        let new_sst = SSTable::write_from_memtable(&dense_idx, new_path(SSTABLES_DIR_PATH))?;
+
+        let old_sst_list: Vec<SSTable>;
+        {
+            // TODO MutexGuard here
+            // In async version, we will have to assume that new sstables may have been created while we were compacting, so we won't be able to just swap.
+            old_sst_list = mem::replace(&mut self.sstables, vec![new_sst]);
```

Comment:
I'm swapping the old list out so that we can read the list and delete the on-disk files.

Reply:
i understand the goal is to keep a handle on the old sstables so they can be subsequently cleaned up. my question regards usage of `mem::replace`. how come it can't be written as:

``` rust
let old_sst_list = self.sstables; // moves vector into local variable
self.sstables = vec![new_sst]; // writes new vec
for old_sst in old_sst_list {
  fs::remove_file(&old_sst.path)?;
}
```

Reply:
The compiler does not allow us to move the thing out of `self.sstables` without assigning it another thing. One option could be to make the type of `self.sstables` be `Option<SSTable>`; then we can `take()` from it and leave it `None`. (In C++ I believe moving out would appear valid, and `self.sstables` would appear to become an empty vector, but a later heap allocation can intrude into it.)

Reply:
I understand now! thanks